{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "# This is a demo file showing the usage of TeloBP, an algorithm for detecting telomere boundary in a DNA sequence.\n",
        "# The algorithm will look for a discontinuity in the telomere pattern, \n",
        "# marking the point where the sequence changes from telomere to non-telomere as the boundary. \n",
        "\n",
        "from Bio import SeqIO\n",
        "import sys\n",
        "\n",
        "sys.path.insert(0, '../TeloBP')\n",
        "from TeloBP import *\n",
        "import constants as c"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "bed_data = []\n",
        "offsetScores = []\n",
        "\n",
        "showGraphs = True\n",
        "\n",
        "# This file is not included in the repository. \n",
        "filename = \"Data/GCA_009914755.4_T2T-CHM13v2.0_genomic.500kb.ends.fna\"\n",
        "for record in SeqIO.parse(filename, \"fasta\"):\n",
        "    chrName = record.id\n",
        "    print(chrName)\n",
        "\n",
        "    if \"p\" in chrName:\n",
        "        distanceFromStart = getTeloBoundary(record.seq[:10000], composition=[[\"CCC\",3/6]], targetPatternIndex=-1, isGStrand=False, teloWindow=100,nucleotideGraphAreaWindowSize=500, windowStep=6,plateauDetectionThreshold=-60, changeThreshold=-20, showGraphs=showGraphs)\n",
        "        print(\"p end: \" + str(distanceFromStart))\n",
        "        pEndIndex = distanceFromStart\n",
        "        bed_data.append([chrName, pEndIndex-6, pEndIndex, chrName])\n",
        "        offset = testTeloLength(chrName, distanceFromStart, c.manualLabelsCHM13)\n",
        "        print(\"offset: \"+str(offset))\n",
        "        offsetScores.append([chrName, offset])\n",
        "    else:\n",
        "        distanceFromEnd = getTeloBoundary(record.seq[-10000:], composition=[[\"GGG\",3/6]], targetPatternIndex=-1, isGStrand = True, teloWindow=100,nucleotideGraphAreaWindowSize=500, windowStep=6,plateauDetectionThreshold=-60, changeThreshold=-20, showGraphs=showGraphs)\n",
        "        print(\"q end: \" + str(distanceFromEnd))\n",
        "        qEndIndex = len(record.seq)-distanceFromEnd\n",
        "        bed_data.append([chrName, qEndIndex, qEndIndex+6, chrName])\n",
        "        offset = testTeloLength(chrName, distanceFromEnd, c.manualLabelsCHM13)\n",
        "        print(\"offset: \"+str(offset))\n",
        "        offsetScores.append([chrName, offset])\n",
        "        \n",
        "# write_bed_file(\"./teloBoundaryOutput/TeloBP%2.0Values.bed\", bed_data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "chr10_MATERNAL\n",
            "chr10_PATERNAL\n",
            "chr11_MATERNAL\n",
            "chr11_PATERNAL\n",
            "chr12_MATERNAL\n",
            "chr12_PATERNAL\n",
            "chr13_MATERNAL\n",
            "chr13_PATERNAL\n",
            "chr14_MATERNAL\n",
            "chr14_PATERNAL\n",
            "chr15_MATERNAL\n",
            "chr15_PATERNAL\n",
            "chr16_MATERNAL\n",
            "chr16_PATERNAL\n",
            "chr17_MATERNAL\n",
            "chr17_PATERNAL\n",
            "chr18_MATERNAL\n",
            "chr18_PATERNAL\n",
            "chr19_MATERNAL\n",
            "chr19_PATERNAL\n",
            "chr1_MATERNAL\n",
            "chr1_PATERNAL\n",
            "chr20_MATERNAL\n",
            "chr20_PATERNAL\n",
            "chr21_MATERNAL\n",
            "chr21_PATERNAL\n",
            "chr22_MATERNAL\n",
            "chr22_PATERNAL\n",
            "chr2_MATERNAL\n",
            "chr2_PATERNAL\n",
            "chr3_MATERNAL\n",
            "chr3_PATERNAL\n",
            "chr4_MATERNAL\n",
            "chr4_PATERNAL\n",
            "chr5_MATERNAL\n",
            "chr5_PATERNAL\n",
            "chr6_MATERNAL\n",
            "chr6_PATERNAL\n",
            "chr7_MATERNAL\n",
            "chr7_PATERNAL\n",
            "chr8_MATERNAL\n",
            "chr8_PATERNAL\n",
            "chr9_MATERNAL\n",
            "chr9_PATERNAL\n",
            "chrEBV\n",
            "chrM\n",
            "chrX_MATERNAL\n",
            "chrY_PATERNAL\n"
          ]
        }
      ],
      "source": [
        "bed_data = []\n",
        "offsetScores = []\n",
        "\n",
        "showGraphs = True\n",
        "\n",
        "# This file is not included in the repository. \n",
        "filename = \"Data/hg002v1.0.1.fasta\"\n",
        "for record in SeqIO.parse(filename, \"fasta\"):\n",
        "    chrName = record.id\n",
        "    print(chrName)\n",
        "    \n",
        "    CStrandTeloLength = getTeloBoundary(record.seq[:500000], isGStrand=False)\n",
        "    GStrandTeloLength = getTeloBoundary(record.seq[-500000:], isGStrand=True)\n",
        "    bed_data.append([chrName, CStrandTeloLength-6, CStrandTeloLength, chrName + \"_CStrand\"])\n",
        "    bed_data.append([chrName, len(record.seq)-GStrandTeloLength, len(record.seq)-GStrandTeloLength+6, chrName + \"_GStrand\"])\n",
        "\n",
        "file = filename.split(\"/\")[-1]\n",
        "\n",
        "write_bed_file(f\"output/TeloBP_{file}.bed\", bed_data)\n",
        "        \n"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "TeloAnalysis",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.11"
    },
    "orig_nbformat": 4
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
